File: 2024/day-07/day7.py
def process_input():
  with open("./day-07/input.txt", "r") as f:
    lines = f.readlines()

    results = []
    components = []
    for line in lines:
      result = line.split()[0][:-1]
      results.append(result)
      curr_components = line.split()[1:]
      components.append(curr_components)

    return results, components

def part1_helper(target_total, curr_total, components, index):
  if curr_total == target_total and index == len(components):
    return True
  if index >= len(components) or curr_total >= target_total:
    return False

  curr_component = int(components[index])
  plus = part1_helper(target_total, curr_total + curr_component, components, index + 1)
  multiply = part1_helper(target_total, curr_total * curr_component, components, index + 1) 
  return plus or multiply

def solve_part1():
  total = 0
  for i in range(len(results)):
    result = int(results[i])
    curr_components = components[i]
    if part1_helper(result, int(curr_components[0]), curr_components, 1):
      # intially did part1_helper(result, 0, curr_components, 1), but this allowed the first 
      # number to be muliplied by 0 which produces an incorrect result.
      total += result
  return total

def part2_helper(target_total, curr_total, components, index):
  if curr_total == target_total and index == len(components):
    return True
  if index >= len(components) or curr_total > target_total:
    return False

  curr_component_string = components[index]
  plus = part2_helper(target_total, curr_total + int(curr_component_string), components, index + 1)
  multiply = part2_helper(target_total, curr_total * int(curr_component_string), components, index + 1)
  # String abuse xd
  concat = part2_helper(target_total, int(str(curr_total) + curr_component_string), components, index + 1) 
  return plus or multiply or concat

def solve_part2():
  total = 0
  for i in range(len(results)):
    result = int(results[i])
    curr_components = components[i]
    if part2_helper(result, int(curr_components[0]), curr_components, 1):
      total += result
  return total

if __name__ == "__main__":
  results, components = process_input()
  print(solve_part1())
  print(solve_part2())

File: 2024/day-07/day7.go
package main

import (
  "fmt"
  "os"
  "strings"
  "strconv"
  "sync"
  "time"
)

func process_input() map[int][]int {
  input, err := os.ReadFile("./day-07/input.txt")
  if err != nil {
    fmt.Println("Error in reading file: ", err)
    os.Exit(1)
  }

  lines := strings.Split(string(input), "\n")

  dict := map[int][]int{}
  for _, line := range lines {
    items := strings.Split(line, " ")

    result, err := strconv.Atoi(items[0][:len(items[0]) - 1])
    if err != nil {
      fmt.Println("Error in converting result to string: ", err)
      os.Exit(1)
    }

    operands := make([]int, len(items[1:]))
    for i, operandStr := range items[1:] {
      operand, err := strconv.Atoi(operandStr)
      if err != nil {
        fmt.Println("Error in converting operand to string: ", err)
        os.Exit(1)
      }
      operands[i] = operand
    }

    dict[result] = operands
  }

  return dict
}

func solve(hasConcat bool) int {
  resultToOperands := process_input()

  total := 0
  var wg sync.WaitGroup
  var mu sync.Mutex
  
  for result, operands := range resultToOperands {
    wg.Add(1)

    go func(result int, operands []int) {
      defer wg.Done()

      if helper(result, operands[0], operands, 1, hasConcat) {
        mu.Lock()
        total += result 
        mu.Unlock()
      }
    }(result, operands)
  }

  wg.Wait()
  return total
}

func helper(target int, curr int, operands []int, i int, hasConcat bool) bool {
  if curr == target && i == len(operands) {
    return true
  }
  if i >= len(operands) || curr > target {
    return false
  }
  operand := operands[i]
  plus := helper(target, curr + operand, operands, i + 1, hasConcat)
  mult := helper(target, curr * operand, operands, i + 1, hasConcat)
  concat := false
  if hasConcat {
    factor := 1
    for operand/factor > 0 {
      factor *= 10
    }
    concat = helper(target, curr * factor + operand, operands, i + 1, hasConcat)
  }
  return plus || mult || concat
}

func main() {
  start := time.Now()
  fmt.Println("Part 1:", solve(false))
  fmt.Println("Part 2:", solve(true))
  fmt.Println("Total time for both parts:", time.Since(start))
}
